fitter/fit_data.py

```diff
@@ -0,0 +1,287 @@
+        Data as a structured array with fields: elevation (deg)
+        plus all fields in COLUMN_NAMES
+    cause : `str`
+        Name of cause being compensated for, e.g. "elevation".
+        Note that `read_data` converts zd to elevation,
+        so you should do the same for this argument.
+    model : callable
+        Model to fit. It must take the following arguments:
+
+        * ang: a numpy.ndarray of angles, in deg
+        * *coeffs: a sequence of coefficients
+    ncoeffs_arr : `list` [`int`]
+        Sequence of numbers of coefficients to try.
+    print_covariance : `bool`, optional
+        Print the covariance matrix from the fit?
+    graph_path : `str` or `None`, optional
+        Path to which to save the graph.
+        If `None` do not save the graph and pause to show it.
+    """
+    cause_data = data["cause"]
+    cause_units = UNITS_DICT[cause]
+    dense_cause_data = np.linspace(start=cause_data[0], stop=cause_data[-1])
+
+    nfits = len(AXES_TO_FIT)
+    fig, plots = plt.subplots(2, nfits, sharex=True, figsize=(20, 10))
+    plot_dict = {axis: (plots[0, i], plots[1, i]) for i, axis in enumerate(AXES_TO_FIT)}
+
+    for axis in AXES_TO_FIT:
+        print()
+        effect_units = UNITS_DICT[axis]
+        effect_data = data[axis]
+
+        data_plot, resid_plot = plot_dict[axis]
+        data_plot.set(ylabel=f"{axis} ({effect_units})")
+        resid_plot.set(
+            xlabel=f"{cause} ({cause_units})", ylabel=f"{axis} fractional residuals)"
+        )
+        data_plot.set_title(f"{axis} vs. {cause}")
+
+        mean_abs_corr = np.mean(np.abs(effect_data))
+        print(f"residuals divisor = {mean_abs_corr:0.3g} (mean(abs({axis} offset))")
+
+        for ncoeffs in ncoeffs_arr:
+            coeffs, covariance = fit_one(
+                data=data, cause=cause, axis=axis, model=model, ncoeffs=ncoeffs
+            )
+            dense_modeled_data = model(dense_cause_data, *coeffs)
+            data_plot.plot(dense_cause_data, dense_modeled_data)
+            residuals = (model(cause_data, *coeffs) - effect_data) / mean_abs_corr
+            resid_plot.plot(cause_data, residuals, ".")
+            print(
+                f"{axis} coeffs: {np.array2string(coeffs, separator=', ', max_line_width=1000)}"
+            )
+            abs_residuals = np.abs(residuals)
+            print(
+                f"{axis} residuals: mean(abs) = {abs_residuals.mean():0.3g}; "
+                f"max(abs) = {abs_residuals.max():0.3g}; "
+                f"std dev = {residuals.std():0.3g}"
+            )
+            if print_covariance:
+                print(f"{axis} coeffs covariance = {covariance}")
+
+        # Plot data last so colors match between data and residual plots.
+        data_plot.plot(cause_data, effect_data, ".")
+
+        legend_labels = [f"{n} coeffs" for n in ncoeffs_arr]
+        resid_plot.legend(legend_labels)
+        fig.show()
+        if graph_path:
+            fig.savefig(graph_path)
```

Comment:
should these be outside of the for loop (for axis in AXES_TO_FIT:)?
we save the figure to disk once after we done looping over dy,dz,and u, right?

Reply:
Thank you for catching that.